src/mod_extdisco.erl

```diff
@@ -0,0 +1,132 @@
+%%==============================================================================
+%% Copyright 2020 Erlang Solutions Ltd.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%% http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%==============================================================================
+-module (mod_extdisco).
+-author('jan.ciesla@erlang-solutions.com').
+
+-xep([{xep, 215}, {version, "0.7"}]).
+-behaviour(gen_mod).
+
+%% gen_mod callbacks.
+-export([start/2, stop/1]).
+
+-export([process_iq/4]).
+
+-include("mongoose.hrl").
+-include("jlib.hrl").
+
+-define(EXT_DISCO, <<"urn:xmpp:extdisco:2">>).
+
+-spec start(jid:server(), list()) -> ok.
+start(Host, Opts) ->
+    IQDisc = gen_mod:get_opt(iqdisc, Opts, one_queue),
```

Comment:
I think selecting `one_queue` option is fine with the functionality :
* no_queue - I think we may occasionally use a queue 
* one_queue - the default one, simple choice 
* {queues, N} - what the N should be? 
* parallel - maybe, but I don't expect extremely high load of those stanzas, so it might be an overkill

I selected more by excluding other options than by a strong grip on any of them.

Reply:
I think that `no_queue` would probably be the best, that means that this iq_stanza is processed in place, synchronously. One queue means that there's a single worker, I understand it as a shortcut for `{queues, 1}`, and then this global process can be a bottleneck. This could make sense for very infrequent requests that actually require a lot of time to compute, that way the c2s process is not stopped waiting for the computations, but this is just querying an ets table, and possibly many many clients could request that arbitrarily.

Reply:
I don't see the gain here in going for `no_queue` as I don't think that processing this IQ is so vital that it needs to be processed immediately.

[https://www.ejabberd.im/files/doc/guide.html#htoc31](https://www.ejabberd.im/files/doc/guide.html#htoc31)
```
* no_queue - All queries of namespace with this processing discipline processed immediately. 
This also means that no other packets can be processed until finished this. 

* one_queue -  In this case created separate queue for processing of IQ queries of namespace with this discipline, 
and processing of this queue is done in parallel with processing of other packets. 
This discipline is most recommended. 

* parallel - In this case for all packets with this discipline spawned separate Erlang process, so all these packets 
processed in parallel. Although spawning of Erlang process have relatively low cost, this can broke server normal
 work, because Erlang emulator have limit on number of processes (32000 by default). 

```

In general we'd expect two such IQs being received at around the same time while two users are establishing a call.

Reply:
It's not about processing it immediately, but rather about the fact that if a hundred users request this information, it will be provided to them one by one, all the requests are processed on a `one_queue` fashion. And I see a possibility of lots of clients wanting to make independent calls, and this is not something that needs to be throttled, so it can be computed in place, on the client's process: https://mongooseim.readthedocs.io/en/latest/developers-guide/Basic-iq-handler/#iq-processing-policies